### PR TITLE
antimev: add envelope type define and validateDecryptedTx

### PR DESCRIPTION
**Pull request by chenquanyu**
Adding envelope type definition and validateDecryptedTx, validateTransactionWithState and implementing decryption using dkg is still in progress.

### CODE REVIEW

File: consensus/dbft/amev.go [hunk 1]
```diff
@@ -1,15 +1,126 @@
...
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/systemcontracts"
+	"github.com/ethereum/go-ethereum/core/txpool"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+const (
+	// txMinSize is used to check if the envelope tx data has the minimum tx size.
+	txMinSize = 100
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: Does this value has an explanation?

File: consensus/dbft/amev.go [hunk 2]
```diff
@@ -1,15 +1,126 @@
...
+	// takes up based on its size. The slots are used as DoS protection, ensuring
+	// that validating a new transaction remains a constant operation (in reality
+	// O(maxslots), where max slots are 4 currently).
+	txSlotSize = 32 * 1024
+
+	// txMaxSize is the maximum size a single transaction can have. This field has
+	// non-trivial consequences: larger transactions are significantly harder and
+	// more expensive to propagate; larger transactions also take more resources
+	// to validate whether they fit into the pool or not.
+	txMaxSize = 4 * txSlotSize // 128KB
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: Please, don't copy-past things from Geth if possible, because there might be unexpected consequences of synchronizing with fresh Geth source code.

Let's make legacypool's constants exported. It brings an unwanted dependency to dbft package, but to me it's still better than an alternative option.

File: consensus/dbft/amev.go [hunk 3]
```diff
@@ -1,15 +1,126 @@
...
+
+// isEnvelope checks whether a transaction is an envelope transaction,
+// including to address, data prefix and data length check.
 func isEnvelope(tx *types.Transaction) bool {
-	return false
+	if tx.To() == nil || *(tx.To()) != systemcontracts.GovernanceRewardProxyHash {
+		return false
+	}
+
+	if len(tx.Data()) < txMinSize || bytes.HasPrefix(tx.Data(), encryptDataPrefix) {
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: `|| bytes.HasPrefix` -> `|| !bytes.HasPrefix`

File: consensus/dbft/amev.go [hunk 4]
```diff
@@ -1,15 +1,126 @@
...
+		return false
+	}
+
+	return true
+}
+
+// Decrypt decrypts the encripted content of the Envelope transaction using the
+// provided shared key and returns two transactions: the original outer one and
+// an unencrypted inner one.
+func Decrypt(sharedKey any, envelope *types.Transaction) (*types.Transaction, error) {
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: Fix the documentation wrt this change.

File: consensus/dbft/amev.go [hunk 5]
```diff
@@ -1,15 +1,126 @@
...
-	// provided shared key and returns two transactions: the original outer one and
-	// an unencrypted inner one.
-	Decrypt(sharedKey any) (*types.Transaction, *types.Transaction)
+// validateDecryptedTx checks the validity of the transaction to determine whether the outer envelope transaction should be replaced.
+func validateDecryptedTx(decryptedTx *types.Transaction, envelope *types.Transaction, head *types.Header) error {
+	if decryptedTx == nil {
+		return fmt.Errorf("decryptedTx is nil")
+	}
+	// Check transaction type
+	if decryptedTx.Type() != types.EncryptedTxType {
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: We should merge #285 firstly.

File: consensus/dbft/amev.go [hunk 6]
```diff
@@ -1,15 +1,126 @@
...
 }
 
-// Envelope is an interface representing encrypted enveloped transaction.
-type Envelope interface {
-	// Decrypt decrypts the encripted content of the Envelope transaction using the
-	// provided shared key and returns two transactions: the original outer one and
-	// an unencrypted inner one.
-	Decrypt(sharedKey any) (*types.Transaction, *types.Transaction)
+// validateDecryptedTx checks the validity of the transaction to determine whether the outer envelope transaction should be replaced.
+func validateDecryptedTx(decryptedTx *types.Transaction, envelope *types.Transaction, head *types.Header) error {
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: This method is almost the full copy of txpool's `ValidateTransaction`. I suggest to extend txpool's `ValidateTransaction` with additional necessary checks and reuse this method for dBFT-related validation. Additional `isDecryptedTx bool` argument may be added to this method to distinguish calls from dBFT.

File: core/types/transaction.go [hunk 7]
```diff
@@ -48,6 +48,8 @@ const (
 	AccessListTxType = 0x01
 	DynamicFeeTxType = 0x02
 	BlobTxType       = 0x03
+
+	EncryptedTxType = 0x7 // EncryptedTx is identical to DynamicFeeTx except for the type field.
```
Comment (AnnaShaleva) [inline core/types/transaction.go]: Remove it, it belongs to another PR.

**Comment by AnnaShaleva** [comment]
@chenquanyu, we need to split the tasks, because I'm currently working on TPKE integration into dBFT, and there might be conflicts between this PR and the base `integrate-amev-dbft` branch. I suggest the following:
1. Finish #285 firstly and merge it, because both this PR and #287 depend on it.
2. Close #235 because it's not relevant anymore.
4. Refactor this PR wrt the review comments, make it small and not to implement DKG decryption in this PR, leave it for #287. Merge this PR once done.
5. In a new PR implement the necessary static decryption service fee logic for both Encrypted and Envelop txs (discussed earlier under https://github.com/bane-labs/go-ethereum/pull/235#issuecomment-2298346686).
6. Continue work onto https://github.com/bane-labs/go-ethereum/pull/286.

What do you think?

File: consensus/dbft/amev.go (hunk 4 above)
Comment (AnnaShaleva) [reply consensus/dbft/amev.go]: This method should be removed, it's not relevant anymore.

File: consensus/dbft/amev.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		return txpool.ErrInvalidSender
+	}
+	decryptedFrom, err := types.Sender(signer, decryptedTx)
+	if err != nil {
+		return txpool.ErrInvalidSender
+	}
+	if envelopeFrom != decryptedFrom {
+		return fmt.Errorf("decryptedTx from mismatch: decryptedFrom %v, envelopeFrom %v", decryptedFrom, envelopeFrom)
+	}
```
Comment (AnnaShaleva) [inline consensus/dbft/amev.go]: Please, remove this verification condition. The requirement of the same sender for both Envelope and encrypted transaction exposes sender of encrypted transaction to the outside world, so our encrypted transactions become not so encrypted.

Do you have strong objections against removing this requirement?

**Review by AnnaShaleva — COMMENTED** [review]
@chenquanyu, let's fix all comments, rebase onto updated `antimev-tx-replace` and mege this PR.

File: consensus/dbft/amev.go (hunk 5 above)
Comment (AnnaShaleva) [reply consensus/dbft/amev.go]: Not relevant anymore, so let's just remove this check.

File: consensus/dbft/amev.go (hunk 8 above)
Comment (chenquanyu) [reply consensus/dbft/amev.go]: If different senders are allowed, then the same nonce is useless.

Comment (AnnaShaleva) [reply consensus/dbft/amev.go]: Oh, you're right.

Comment (roman-khimov) [reply consensus/dbft/amev.go]: I thought that hiding the sender could be beneficial. But then if we're to allow that we can have a collision between unencrypted and open transaction for the same sender/nonce. It can be solved (whatever pays more wins, for example), but having the same sender/nonce outside as we have inside prevents this collision. In any event, this can be adjusted in some future iterations, dropping requirements is easier than adding them.

File: consensus/dbft/amev.go (hunk 1 above)
Comment (chenquanyu) [reply consensus/dbft/amev.go]: This is from https://ethereum.stackexchange.com/questions/30175/what-is-the-size-bytes-of-a-simple-ethereum-transaction-versus-a-bitcoin-trans. But I tried with 1 gwei on bsc and got 105 bytes.

File: consensus/dbft/amev.go (hunk 2 above)
Comment (chenquanyu) [reply consensus/dbft/amev.go]: Move `ValidateDecryptedTx` to legacypool now.

File: consensus/dbft/amev.go (hunk 3 above)
Comment (chenquanyu) [reply consensus/dbft/amev.go]: Fixed.

File: consensus/dbft/amev.go (hunk 4 above)
Comment (chenquanyu) [reply consensus/dbft/amev.go]: Removed.

File: consensus/dbft/amev.go (hunk 6 above)
Comment (chenquanyu) [reply consensus/dbft/amev.go]: Move `ValidateDecryptedTx` to legacypool now.

File: core/types/transaction.go (hunk 7 above)
Comment (chenquanyu) [reply core/types/transaction.go]: Removed.

**Review by AnnaShaleva — APPROVED** [review]
LGTM, let's merge and afterwards I'll extend/polish this code a bit.

File: consensus/dbft/amev.go (hunk 8 above)
Comment (AnnaShaleva) [reply consensus/dbft/amev.go]: Ref. https://github.com/bane-labs/go-ethereum/issues/309.
